addons/purchase_stock/models/res_partner.py

```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from datetime import timedelta
+
+from odoo import fields, models
+
+
+class res_partner(models.Model):
+    _inherit = 'res.partner'
+
+    def _compute_on_time_rate(self):
+        for partner in self:
+            order_lines = partner.env['purchase.order'].search([
+                ('partner_id', '=', partner.id),
+                ('date_order', '>', fields.Date.today() - timedelta(365)),
+                ('state', 'in', ['done', 'purchase']),
+            ]).mapped('order_line').filtered(lambda l: l.move_ids.quantity_done)    # may use other to decide
```

Comment:
this seems pretty costly + you have a stored partner_id directly on po line, you should try to do a read_group out of the loop using it

Reply:
I thought normally we won't have the situation many partner changed on batch, But ok, I will change it.

Reply:
@sle-odoo I can't use read_group, since I need the date, number not to be aggregated, but to reduce the query numbers, I can search all POL and group them manually.